Tests/SourceKitDTests/CrashRecoveryTests.swift

```diff
@@ -0,0 +1,276 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import ISDBTestSupport
+import LanguageServerProtocol
+import SourceKitLSP
+import SourceKitD
+import SKTestSupport
+import XCTest
+
+final class CrashRecoveryTests: XCTestCase {
+  func testSourcekitdCrashRecovery() {
+    guard longTestsEnabled else {
+      return
+    }
+
+    let ws = try! staticSourceKitTibsWorkspace(name: "sourcekitdCrashRecovery")!
+    let loc = ws.testLoc("loc")
+
+    try! ws.openDocument(loc.url, language: .swift)
+
+    // Make a change to the file that's not saved to disk. This way we can check that we re-open the correct in-memory state.
+
+    let addFuncChange = TextDocumentContentChangeEvent(range: loc.position..<loc.position, rangeLength: 0, text: """
+
+      func foo() {
+        print("Hello world")
+      }
+      """)
+    ws.sk.send(DidChangeTextDocumentNotification(textDocument: VersionedTextDocumentIdentifier(loc.docUri, version: 2), contentChanges: [addFuncChange]))
+
+    // Do a sanity check and verify that we get the expected result from a hover response before crashing sourcekitd.
+
+    let expectedMarkupContent = MarkupContent(kind: .markdown, value: """
```

Comment:
Yeah, you're right. I suppose, I was lazy because getting a string value of the response is non-trivial because of all the different representations the response can take. I've added a `file private` extension in the test to check if a `HoverReponse` contains a substring.